### PR TITLE
Epmlstrcmw 122: microtypes for user

### PR DESCRIPTION
**Pull request by Dmitrii-Shevchenko**
Refine domain-level types and related logic with microtypes

### CODE REVIEW

File: controllers/src/test/scala/cromwell/pipeline/controller/AuthControllerTest.scala [hunk 1]
```diff
@@ -20,31 +22,28 @@ class AuthControllerTest extends WordSpec with Matchers with MockFactory with Sc
...
-      "return token headers if user exists" taggedAs Controller in {
-        val email = "email@cromwell.com"
-        val password = "password"
+      "return token headers if user exists" in {
+        val signInRequest = SignInRequest(UserEmail("email@cromwell.com"), password)
         val authResponse = AuthResponse(accessToken, refreshToken, accessTokenExpiration)
-        val signInRequestStr = s"""{"email":"${email}","password":"${password}"}"""
-        val httpEntity = HttpEntity(`application/json`, signInRequestStr)
-        (authService.signIn _ when SignInRequest(email, password)).returns(Future(Option(authResponse)))
+        val httpEntity = HttpEntity(`application/json`, Json.stringify(Json.toJson(signInRequest)))
```
Comment (OstapStrashevskii) [inline controllers/src/test/scala/cromwell/pipeline/controller/AuthControllerTest.scala]: don't care of this comment, but why do you use "stringify" here?

File: controllers/src/test/scala/cromwell/pipeline/controller/AuthControllerTest.scala [hunk 2]
```diff
@@ -20,31 +22,28 @@ class AuthControllerTest extends WordSpec with Matchers with MockFactory with Sc
...
+        (authService.signIn _ when signInRequest).returns(Future(Option(authResponse)))
```
Comment (OstapStrashevskii) [inline controllers/src/test/scala/cromwell/pipeline/controller/AuthControllerTest.scala]: maybe Future.successfull is relevant here

File: controllers/src/test/scala/cromwell/pipeline/controller/AuthControllerTest.scala [hunk 3]
```diff
@@ -54,56 +53,56 @@ class AuthControllerTest extends WordSpec with Matchers with MockFactory with Sc
...
+          Name("LastName")
+        )
+        val signUpJson = Json.stringify(Json.toJson(signUpRequest))
         val authResponse = AuthResponse(accessToken, refreshToken, accessTokenExpiration)
 
-        val httpEntity = HttpEntity(`application/json`, signUpRequestStr)
-        (authService.signUp _ when SignUpRequest(email, password, firstName, lastName))
-          .returns(Future(Some(authResponse)))
+        val httpEntity = HttpEntity(`application/json`, signUpJson)
+        (authService.signUp _ when signUpRequest).returns(Future(Some(authResponse)))
```
Comment (OstapStrashevskii) [inline controllers/src/test/scala/cromwell/pipeline/controller/AuthControllerTest.scala]: the same thing(Future)

File: services/src/test/scala/cromwell/pipeline/service/UserServiceTest.scala [hunk 4]
```diff
@@ -55,35 +46,43 @@ class UserServiceTest extends AsyncWordSpec with Matchers with MockitoSugar {
...
+          firstName = Name("updatedFirstName"),
+          lastName = Name("updatedLastName")
+        )
+        val request = UserUpdateRequest(
+          user2.email,
+          user2.firstName,
+          user2.lastName
+        )
+
+        when(userRepository.getUserById(user.userId)).thenReturn(Future(Some(user)))
```
Comment (OstapStrashevskii) [inline services/src/test/scala/cromwell/pipeline/service/UserServiceTest.scala]: Future.successful ?

File: repositories/src/main/scala/cromwell/pipeline/datastorage/dto/User.scala [hunk 5]
```diff
@@ -1,35 +1,76 @@
...
+final class UserEmail private (override val unwrap: String) extends AnyVal with Wrapped[String]
+
+object UserEmail extends Wrapped.Companion {
+  type Type = String
+  type Wrapper = UserEmail
+  type Error = String
+  implicit lazy val emailFormat: Format[UserEmail] = Json.valueFormat[UserEmail]
+  override protected def create(value: String): UserEmail = new UserEmail(value)
+  override protected def validate(value: String): ValidationResult[String] = Validated.cond(
+    value.matches("^[^@]+@[^\\.]+\\..+$"),
```
Comment (GrigoriiBerezin) [inline repositories/src/main/scala/cromwell/pipeline/datastorage/dto/User.scala]: Is text1@text2@.text3@ valid email?

File: services/src/test/scala/cromwell/pipeline/service/AuthServiceTest.scala [hunk 6]
```diff
@@ -3,7 +3,7 @@ package cromwell.pipeline.service
 import java.time.Instant
 
 import cromwell.pipeline.datastorage.dao.repository.UserRepository
-import cromwell.pipeline.datastorage.dto.UserId
+import cromwell.pipeline.datastorage.dto.{ UUID }
```
Comment (GrigoriiBerezin) [inline services/src/test/scala/cromwell/pipeline/service/AuthServiceTest.scala]: Brace can be omitted

File: logs/log-20200318T222651.txt [hunk 7]
```diff
@@ -0,0 +1,2640 @@
+555  [main] WARN  com.zaxxer.hikari.HikariConfig - database.postgres_dc.db - using dataSourceClassName and ignoring jdbcUrl.
```
Comment (GrigoriiBerezin) [inline logs/log-20200318T222651.txt]: I think the logs should be removed from PR

File: logs/log-20200323T193840.txt [hunk 8]
```diff
@@ -0,0 +1,2643 @@
+511  [main] WARN  com.zaxxer.hikari.HikariConfig - database.postgres_dc.db - using dataSourceClassName and ignoring jdbcUrl.
```
Comment (nikita0levchenko) [inline logs/log-20200323T193840.txt]: I think, we don't need log's in this PR

File: services/src/main/scala/cromwell/pipeline/service/UserService.scala [hunk 9]
```diff
@@ -13,38 +13,45 @@ class UserService(userRepository: UserRepository)(implicit executionContext: Exe
...
-      } match {
-        case Some(value) => value
-        case None        => Future.failed(new RuntimeException("user with this id doesn't exist"))
-      }
-    }
+  def updateUser(userId: UUID, request: UserUpdateRequest): Future[Int] =
+    userRepository
+      .getUserById(userId)
+      .flatMap(
+        userOpt =>
```
Comment (nikita0levchenko) [inline services/src/main/scala/cromwell/pipeline/service/UserService.scala]: I think, we can use pattern matching on the userOpt. I did the same thing on my updateProject method

**Comment by ilyaonishenko** [comment]
Please fix a typo in the MR name..

File: controllers/src/test/scala/cromwell/pipeline/controller/AuthControllerTest.scala (hunk 2 above)
Comment (Dmitrii-Shevchenko) [reply controllers/src/test/scala/cromwell/pipeline/controller/AuthControllerTest.scala]: I don't think so, because signIn method expects an Option type

File: controllers/src/test/scala/cromwell/pipeline/controller/AuthControllerTest.scala (hunk 3 above)
Comment (Dmitrii-Shevchenko) [reply controllers/src/test/scala/cromwell/pipeline/controller/AuthControllerTest.scala]: the same thing

File: services/src/test/scala/cromwell/pipeline/service/UserServiceTest.scala (hunk 4 above)
Comment (Dmitrii-Shevchenko) [reply services/src/test/scala/cromwell/pipeline/service/UserServiceTest.scala]: Nope

File: controllers/src/test/scala/cromwell/pipeline/controller/AuthControllerTest.scala (hunk 1 above)
Comment (Dmitrii-Shevchenko) [reply controllers/src/test/scala/cromwell/pipeline/controller/AuthControllerTest.scala]: Done

File: services/src/test/scala/cromwell/pipeline/service/AuthServiceTest.scala (hunk 6 above)
Comment (Dmitrii-Shevchenko) [reply services/src/test/scala/cromwell/pipeline/service/AuthServiceTest.scala]: Done

File: logs/log-20200318T222651.txt (hunk 7 above)
Comment (Dmitrii-Shevchenko) [reply logs/log-20200318T222651.txt]: Done

File: logs/log-20200323T193840.txt (hunk 8 above)
Comment (Dmitrii-Shevchenko) [reply logs/log-20200323T193840.txt]: Done

File: services/src/main/scala/cromwell/pipeline/service/UserService.scala (hunk 9 above)
Comment (Dmitrii-Shevchenko) [reply services/src/main/scala/cromwell/pipeline/service/UserService.scala]: Done

File: services/src/test/scala/cromwell/pipeline/service/UserServiceTest.scala (hunk 4 above)
Comment (myazinn) [reply services/src/test/scala/cromwell/pipeline/service/UserServiceTest.scala]: Why not?

File: repositories/src/test/scala/cromwell/pipeline/datastorage/dao/repository/utils/TestUserUtils.scala [hunk 10]
```diff
@@ -1,27 +1,25 @@
...
+    firstName: Name = Name("FirstName"),
+    lastName: Name = Name("LastName"),
     active: Boolean = true
   ): User = {
     val passwordHash = StringUtils.calculatePasswordHash(password, passwordSalt)
     User(
-      UserId(uuid),
-      s"JohnDoe-$uuid@cromwell.com",
+      uuid,
+      UserEmail(s"JohnDoe-${uuid.unwrap}@cromwell.com"),
```
Comment (myazinn) [inline repositories/src/test/scala/cromwell/pipeline/datastorage/dao/repository/utils/TestUserUtils.scala]: Something is wrong.
We should not be able to unsafely create an instance of `UserEmail` without `Unsafe` object.
@DenisVerkhoturov what is going on? :)

File: repositories/src/main/scala/cromwell/pipeline/datastorage/dto/User.scala (hunk 5 above)
Comment (Dmitrii-Shevchenko) [reply repositories/src/main/scala/cromwell/pipeline/datastorage/dto/User.scala]: I think there is no perfect regex for email, best way to approve email is sending to users an activation link imho, but if you have this one share please

File: repositories/src/test/scala/cromwell/pipeline/datastorage/dao/repository/utils/TestUserUtils.scala (hunk 10 above)
Comment (Dmitrii-Shevchenko) [reply repositories/src/test/scala/cromwell/pipeline/datastorage/dao/repository/utils/TestUserUtils.scala]: My fault, fixed

File: services/src/test/scala/cromwell/pipeline/service/UserServiceTest.scala (hunk 4 above)
Comment (Dmitrii-Shevchenko) [reply services/src/test/scala/cromwell/pipeline/service/UserServiceTest.scala]: Because method getUserById expects an Future[Option[User]] type, how to do otherwise?

Comment (Dmitrii-Shevchenko) [reply services/src/test/scala/cromwell/pipeline/service/UserServiceTest.scala]: `Future.successful(Some(user)) `
This is the way?

Comment (myazinn) [reply services/src/test/scala/cromwell/pipeline/service/UserServiceTest.scala]: This is the way

Comment (Dmitrii-Shevchenko) [reply services/src/test/scala/cromwell/pipeline/service/UserServiceTest.scala]: Done

File: repositories/src/test/scala/cromwell/pipeline/datastorage/dao/repository/utils/TestUserUtils.scala (hunk 10 above)
Comment (Dmitrii-Shevchenko) [reply repositories/src/test/scala/cromwell/pipeline/datastorage/dao/repository/utils/TestUserUtils.scala]: Done

File: repositories/src/main/scala/cromwell/pipeline/datastorage/dto/User.scala [hunk 11]
```diff
@@ -1,35 +1,79 @@
...
+  )
+}
+
+final class Name private (override val unwrap: String) extends AnyVal with Wrapped[String]
+
+object Name extends Wrapped.Companion {
+  type Type = String
+  type Wrapper = Name
+  type Error = String
+  implicit lazy val nameFormat: Format[Name] = Json.valueFormat[Name]
```
Comment (Dmitrii-Shevchenko) [inline repositories/src/main/scala/cromwell/pipeline/datastorage/dto/User.scala:53]: how to do suitable formatter which couldn't ignore validation?

File: logs/log-20200318T222651.txt (hunk 7 above)
Comment (mnesvit) [reply logs/log-20200318T222651.txt:1]: @Dmitrii-Shevchenko  I still see this files in list of changes in PR.

Comment (Dmitrii-Shevchenko) [reply logs/log-20200318T222651.txt]: Fixed, thank you!

File: model/src/main/scala/cromwell/pipeline/model/wrapper/Name.scala [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+final class Name private (override val unwrap: String) extends AnyVal with Wrapped[String]
+
+object Name extends Wrapped.Companion {
+  type Type = String
+  type Wrapper = Name
+  type Error = String
+  implicit lazy val nameFormat: Format[Name] = new Format[Name] {
+    override def reads(json: JsValue): JsResult[Name] = Name.from(json.as[String]) match {
+      case Valid(x)   => JsSuccess(Name(x))
+      case Invalid(e) => JsError(e.head)
```
Comment (myazinn) [inline model/src/main/scala/cromwell/pipeline/model/wrapper/Name.scala:18]: I think we must keep all errors, not only the first error.

File: portal/src/main/scala/cromwell/pipeline/CromwellPipelineApp.scala [hunk 13]
```diff
@@ -32,6 +34,14 @@ object CromwellPipelineApp extends App {
...
   }
 
+  implicit def rejectionHandler =
+    RejectionHandler.default.mapRejectionResponse {
+      case res @ HttpResponse(_, _, ent: HttpEntity.Strict, _) =>
+        val message = ent.data.utf8String.replaceAll("\"", """\"""")
+        res.withEntity(HttpEntity(ContentTypes.`application/json`, message))
+      case x => x
+    }
+
```
Comment (myazinn) [inline portal/src/main/scala/cromwell/pipeline/CromwellPipelineApp.scala:44]: This rejection handler basically does nothing. This is not what we discussed.

File: repositories/src/main/scala/cromwell/pipeline/datastorage/utils/auth/SecurityDirective.scala [hunk 14]
```diff
@@ -25,9 +25,8 @@ class SecurityDirective(authConfig: AuthConfig) {
           .map(provide)
           .getOrElse(complete(StatusCodes.Unauthorized -> UnauthorizedMessages.AnotherTypeOfToken))
       case None =>
-        complete(StatusCodes.Unauthorized -> UnauthorizedMessages.MissedToken)
+        reject()
```
Comment (myazinn) [inline repositories/src/main/scala/cromwell/pipeline/datastorage/utils/auth/SecurityDirective.scala:28]: We must provide custom rejection here, don't we?

File: services/src/test/scala/cromwell/pipeline/service/AuthServiceTest.scala [hunk 15]
```diff
@@ -30,25 +30,23 @@ class AuthServiceTest extends WordSpec with Matchers with MockFactory {
...
 
-      "return Auth response for active refresh token" taggedAs Service in new RefreshTokenContext(
-        expirationTimeInSeconds.refreshToken
-      ) {
+      "return Auth response for active refresh token" in new RefreshTokenContext(expirationTimeInSeconds.refreshToken) {
         authService.refreshTokens(refreshToken) shouldBe Some(authResponse)
       }
 
-      "return None for outdated refresh token" taggedAs Service in new RefreshTokenContext(lifetime = 0) {
+      "return None for outdated refresh token" in new RefreshTokenContext(lifetime = 0) {
```
Comment (myazinn) [inline services/src/test/scala/cromwell/pipeline/service/AuthServiceTest.scala:43]: Same question about removing `taggedAs`

File: controllers/src/test/scala/cromwell/pipeline/controller/UserControllerTest.scala [hunk 16]
```diff
@@ -19,23 +21,23 @@ class UserControllerTest
     with Matchers
     with ScalatestRouteTest
     with MockitoSugar
-    with BeforeAndAfterAll {
-  import PlayJsonSupport._
+    with PlayJsonSupport {
```
Comment (myazinn) [inline controllers/src/test/scala/cromwell/pipeline/controller/UserControllerTest.scala:24]: We discussed that we should import `PlayJsonSupport._` instead of mixing it. Why did you change it here?

File: controllers/src/test/scala/cromwell/pipeline/controller/AuthControllerTest.scala [hunk 17]
```diff
@@ -54,55 +55,38 @@ class AuthControllerTest extends WordSpec with Matchers with MockFactory with Sc
...
-        val email = "email@cromwell.com"
-        val password = "password"
-        val firstName = "First-name"
-        val lastName = "Last-name"
+      "return BadRequest status if user registration was failed" in {
         val signUpRequestStr =
           s"""{"email":"${email}","password":"${password}","firstName":"${firstName}","lastName":"${lastName}"}"""
         val httpEntity = HttpEntity(`application/json`, signUpRequestStr)
-        (authService.signUp _ when SignUpRequest(email, password, firstName, lastName))
+        (authService.signUp _ when SignUpRequest(UserEmail(email), Password(password), Name(firstName), Name(lastName)))
```
Comment (myazinn) [inline controllers/src/test/scala/cromwell/pipeline/controller/AuthControllerTest.scala:89]: For some reason, we are still able to unsafely create microtypes without explicitly importing the `Unsafe` object.
This must be fixed.

File: controllers/src/test/scala/cromwell/pipeline/controller/AuthControllerTest.scala [hunk 18]
```diff
@@ -20,31 +20,32 @@ class AuthControllerTest extends WordSpec with Matchers with MockFactory with Sc
...
         Post("/auth/signIn", httpEntity) ~> authController.route ~> check {
           status shouldBe StatusCodes.OK
           checkAuthTokens
         }
       }
 
-      "return Unauthorized status if user doesn't exist" taggedAs Controller in {
-        val email = "email@cromwell.com"
-        val password = "password"
+      "return Unauthorized status if user doesn't exist" in {
```
[the remark this answers is not in the document]
Comment (Dmitrii-Shevchenko) [reply controllers/src/test/scala/cromwell/pipeline/controller/AuthControllerTest.scala]: Done

File: controllers/src/test/scala/cromwell/pipeline/controller/ProjectControllerTest.scala [hunk 19]
```diff
@@ -39,7 +41,7 @@ class ProjectControllerTest extends AsyncWordSpec with Matchers with ScalatestRo
     "delete project by id" should {
       "return deactivated project entity" in {
         val dummyProject = TestProjectUtils.getDummyProject()
-        val accessToken = AccessTokenContent(TestUserUtils.getDummyUserId.value)
+        val accessToken = AccessTokenContent(TestUserUtils.getDummyUser().userId)
```
[the remark this answers is not in the document]
Comment (Dmitrii-Shevchenko) [reply controllers/src/test/scala/cromwell/pipeline/controller/ProjectControllerTest.scala]: Done

File: controllers/src/test/scala/cromwell/pipeline/controller/ProjectControllerTest.scala [hunk 20]
```diff
@@ -82,8 +84,8 @@ class ProjectControllerTest extends AsyncWordSpec with Matchers with ScalatestRo
       }
 
       "return InternalServerError status if the update failed" in {
-        val userId = TestUserUtils.getDummyUserId
-        val accessToken = AccessTokenContent("0")
+        val userId = TestUserUtils.getDummyUser().userId
+        val accessToken = AccessTokenContent(TestUserUtils.getDummyUser().userId) //was 0
```
[the remark this answers is not in the document]
Comment (Dmitrii-Shevchenko) [reply controllers/src/test/scala/cromwell/pipeline/controller/ProjectControllerTest.scala]: Done

File: services/src/test/scala/cromwell/pipeline/service/AuthServiceTest.scala (hunk 15 above)
Comment (Dmitrii-Shevchenko) [reply services/src/test/scala/cromwell/pipeline/service/AuthServiceTest.scala]: Done

File: controllers/src/test/scala/cromwell/pipeline/controller/UserControllerTest.scala (hunk 16 above)
Comment (Dmitrii-Shevchenko) [reply controllers/src/test/scala/cromwell/pipeline/controller/UserControllerTest.scala]: Done

File: model/src/main/scala/cromwell/pipeline/model/wrapper/Name.scala (hunk 12 above)
Comment (Dmitrii-Shevchenko) [reply model/src/main/scala/cromwell/pipeline/model/wrapper/Name.scala]: Done

File: model/src/main/scala/cromwell/pipeline/model/wrapper/Name.scala [hunk 21]
```diff
@@ -0,0 +1,29 @@
...
+
+final class Name private (override val unwrap: String) extends AnyVal with Wrapped[String]
+
+object Name extends Wrapped.Companion {
+  type Type = String
+  type Wrapper = Name
+  type Error = String
+  implicit lazy val nameFormat: Format[Name] = new Format[Name] {
+    override def reads(json: JsValue): JsResult[Name] = Name.from(json.as[String]) match {
+      case Valid(x)   => JsSuccess(Name(x))
```
[the remark this answers is not in the document]
Comment (Dmitrii-Shevchenko) [reply model/src/main/scala/cromwell/pipeline/model/wrapper/Name.scala]: Done

File: repositories/src/main/scala/cromwell/pipeline/datastorage/utils/auth/SecurityDirective.scala (hunk 14 above)
Comment (Dmitrii-Shevchenko) [reply repositories/src/main/scala/cromwell/pipeline/datastorage/utils/auth/SecurityDirective.scala]: Done

File: portal/src/main/scala/cromwell/pipeline/CromwellPipelineApp.scala (hunk 13 above)
Comment (Dmitrii-Shevchenko) [reply portal/src/main/scala/cromwell/pipeline/CromwellPipelineApp.scala]: Done

File: model/src/main/scala/cromwell/pipeline/model/validator/Wrapped.scala [hunk 22]
```diff
@@ -0,0 +1,43 @@
...
+  override def toString: String = unwrap.toString
+}
+object Wrapped {
+  trait Companion {
+    type Type
+    type Error
+    type Wrapper <: Wrapped[Type]
+    type ValidationResult[A] = Validated[NonEmptyChain[Error], A]
+    implicit def wrappedOrdering(implicit ord: Ordering[Type]): Ordering[Wrapper] = Ordering.by(_.unwrap)
+    implicit def unwrap(wrapped: Wrapper): Type = wrapped.unwrap
```
[the remark this answers is not in the document]
Comment (Dmitrii-Shevchenko) [reply model/src/main/scala/cromwell/pipeline/model/validator/Wrapped.scala]: Done

File: controllers/src/test/scala/cromwell/pipeline/controller/AuthControllerTest.scala (hunk 17 above)
Comment (Dmitrii-Shevchenko) [reply controllers/src/test/scala/cromwell/pipeline/controller/AuthControllerTest.scala]: Done

File: portal/src/it/scala/cromwell/pipeline/controller/ProjectControllerItTest.scala [hunk 23]
```diff
@@ -36,15 +37,16 @@ class ProjectControllerItTest
   "ProjectController" when {
     "getProjectByName" should {
       "return a project with the same name" in {
-        val dummyProject = TestProjectUtils.getDummyProject(ownerId = dummyUser.userId)
+        val dummyProject =
+          TestProjectUtils.getDummyProject(ProjectId(UserId.random.unwrap), dummyUser.userId)
```
Comment (myazinn) [inline portal/src/it/scala/cromwell/pipeline/controller/ProjectControllerItTest.scala:41]: `ProjectId` is created using `UserId`?:) It looks strange.

File: repositories/src/test/scala/cromwell/pipeline/datastorage/dao/repository/utils/TestProjectUtils.scala [hunk 24]
```diff
@@ -1,19 +1,18 @@
...
 
-import java.util.UUID
-
-import cromwell.pipeline.datastorage.dto._
+import cromwell.pipeline.datastorage.dto.{ Project, ProjectId }
+import cromwell.pipeline.model.wrapper.UserId
 
 object TestProjectUtils {
+  val userUuid = UserId.random
+  val projectUuid = java.util.UUID.randomUUID().toString
```
Comment (myazinn) [inline repositories/src/test/scala/cromwell/pipeline/datastorage/dao/repository/utils/TestProjectUtils.scala:8]: Why not keep them `def` instead of `val`? You won't have to provide custom `ProjectId` everywhere in tests.

File: utils/src/main/scala/cromwell/pipeline/utils/AccessTokenRejection.scala [hunk 25]
```diff
@@ -0,0 +1,5 @@
+package cromwell.pipeline.utils
+
+import akka.http.scaladsl.server.Rejection
+
+final case class AccessTokenRejection(msg: String) extends Rejection
```
Comment (myazinn) [inline utils/src/main/scala/cromwell/pipeline/utils/AccessTokenRejection.scala:5]: Maybe `MissingAccessTokenRejection` ?

Comment (Dmitrii-Shevchenko) [reply utils/src/main/scala/cromwell/pipeline/utils/AccessTokenRejection.scala]: Done

File: repositories/src/test/scala/cromwell/pipeline/datastorage/dao/repository/utils/TestProjectUtils.scala (hunk 24 above)
Comment (Dmitrii-Shevchenko) [reply repositories/src/test/scala/cromwell/pipeline/datastorage/dao/repository/utils/TestProjectUtils.scala]: Done

File: portal/src/it/scala/cromwell/pipeline/controller/ProjectControllerItTest.scala (hunk 23 above)
Comment (Dmitrii-Shevchenko) [reply portal/src/it/scala/cromwell/pipeline/controller/ProjectControllerItTest.scala]: Done

File: model/src/main/scala/cromwell/pipeline/model/validator/Wrapped.scala [hunk 26]
```diff
@@ -0,0 +1,54 @@
...
+  override def toString: String = unwrap.toString
+}
+object Wrapped {
+  trait Companion {
+    type Type
+    type Error
+    type Wrapper <: Wrapped[Type]
+    type ValidationResult[A] = Validated[NonEmptyChain[Error], A]
+    implicit def wrappedOrdering(implicit ord: Ordering[Type]): Ordering[Wrapper] = Ordering.by(_.unwrap)
+    def unwrap(wrapped: Wrapper): Type = wrapped.unwrap
```
[the remark this answers is not in the document]
Comment (Dmitrii-Shevchenko) [reply model/src/main/scala/cromwell/pipeline/model/validator/Wrapped.scala]: Done

File: controllers/src/main/scala/cromwell/pipeline/controller/ProjectController.scala [hunk 27]
```diff
@@ -31,7 +34,7 @@ class ProjectController(projectService: ProjectService)(
         },
         post {
           entity(as[ProjectAdditionRequest]) { request =>
-            onComplete(projectService.addProject(request, UserId(accessToken.userId), "repoStub")) {
+            onComplete(projectService.addProject(request, UserId(accessToken.userId.unwrap, Enable.Unsafe), "repoStub")) {
```
[the remark this answers is not in the document]
Comment (Dmitrii-Shevchenko) [reply controllers/src/main/scala/cromwell/pipeline/controller/ProjectController.scala]: Done
